packages/models-library/src/models_library/user_preferences.py

```diff
@@ -0,0 +1,186 @@
+from collections.abc import Mapping
+from enum import Enum, auto
+from typing import Any, ClassVar, TypeAlias
+
+from pydantic import BaseModel, Field
+from pydantic.main import ModelMetaclass
+
+from .services import ServiceKey
+from .services_ui import WidgetType
+from .utils.enums import StrAutoEnum
+
+IntStr: TypeAlias = int | str
+DictStrAny = dict[str, Any]
+AbstractSetIntStr: TypeAlias = set[IntStr]
+MappingIntStrAny: TypeAlias = Mapping[IntStr, Any]
+
+
+class _AutoRegisterMeta(ModelMetaclass):
+    _registered_user_preference_classes: ClassVar[dict[str, type]] = {}
+
+    def __new__(cls, name, bases, attrs):
+        new_class = super().__new__(cls, name, bases, attrs)
+
+        if name != BaseModel.__name__:
+            if name in cls._registered_user_preference_classes:
+                msg = (
+                    f"Class named '{name}' was already defined at "
+                    f"{cls._registered_user_preference_classes[name]}."
+                    " Please choose a different class name!"
+                )
+                raise TypeError(msg)
+            cls._registered_user_preference_classes[name] = new_class
+
+        return new_class
+
+
+PreferenceName: TypeAlias = str
+
+
+class _ExtendedBaseModel(BaseModel, metaclass=_AutoRegisterMeta):
+    ...
+
+
+class ValueType(str, Enum):
+    BOOL = "boolean"
+    STR = "string"
+    FLOAT = "number"
+    INT = "integer"
+    LIST = "array"
+    DICT = "dictionary"
+
+
+class PreferenceType(StrAutoEnum):
+    BACKEND = auto()
+    FRONTEND = auto()
+    USER_SERVICE = auto()
+
+
+class NoPreferenceFoundError(RuntimeError):
+    ...
+
+
+class _BaseUserPreferenceModel(_ExtendedBaseModel):
+    preference_type: PreferenceType = Field(
+        ..., description="distinguish between the types of preferences"
+    )
+
+    value: Any = Field(
+        ...,
+        description="the value of the preference. Stored as is and cannot be queried over",
+    )
+
+    @classmethod
+    def get_preference_class_from_name(
+        cls, preference_name: str
+    ) -> "_BaseUserPreferenceModel":
+        preference_class: "_BaseUserPreferenceModel" | None = (
```

Comment:
As I mention above, please consider building class models using `pydantic.create_model` from value stored in a global registry  (i.e. name-fields)

Reply:
Maybe there was a bit of confusion here. The typing was wrong. I just return the class name not an object.

And to clarify you comment, the model already exists. Nothing is created at runtime.